Source/WebKit/NetworkProcess/webrtc/rice/RiceBackend.cpp

```diff
@@ -0,0 +1,359 @@
+}
+
+void recvSourceWakeup(GSource* base)
+{
+    auto source = reinterpret_cast<RecvSource*>(base);
+    auto context = g_source_get_context(base);
+    source->needsDispatch.exchange(1);
+
+    if (context)
+        g_main_context_wakeup(context);
+}
+
+static void recvSourceFinalize(GSource*)
+{
+}
+
+static GSourceFuncs recvSourceEventFunctions = {
+    recvSourcePrepare,
+    recvSourceCheck,
+    recvSourceDispatch,
+    recvSourceFinalize,
+    nullptr, nullptr
+};
+
+static GRefPtr<GSource> recvSourceNew()
+{
+    auto source = adoptGRef(g_source_new(&recvSourceEventFunctions, sizeof(RecvSource)));
+    g_source_set_priority(source.get(), RunLoopSourcePriority::AsyncIONetwork);
+    g_source_set_name(source.get(), "[WebKit] ICE Agent recv loop");
+
+    auto recvSource = reinterpret_cast<RecvSource*>(source.get());
+    recvSource->needsDispatch.exchange(1);
+
+    return source;
+}
+
+struct RecvSourceData {
+    RiceBackend* backend;
+    unsigned streamId;
+};
+WEBKIT_DEFINE_ASYNC_DATA_STRUCT(RecvSourceData);
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(RiceBackend);
+
+void RiceBackend::initialize(NetworkConnectionToWebProcess& connectionToWebProcess, WebKit::WebPageProxyIdentifier&&, CompletionHandler<void(RefPtr<RiceBackend>&&)>&& completionHandler)
+{
+    Ref backend = RiceBackend::create(connectionToWebProcess);
+    completionHandler(WTFMove(backend));
+}
+
+RiceBackend::RiceBackend(NetworkConnectionToWebProcess& connection)
+    : m_connection(connection)
+{
+    static Atomic<uint32_t> counter = 0;
+    auto id = counter.load();
+    auto threadName = makeString("webrtc-rice-"_s, id);
+    counter.exchangeAdd(1);
+
+    m_runLoop = RunLoop::create(ASCIILiteral::fromLiteralUnsafe(threadName.ascii().data()));
+}
+
+RiceBackend::~RiceBackend() = default;
+
+IPC::Connection* RiceBackend::messageSenderConnection() const
+{
+    return m_connection ? &m_connection->connection() : nullptr;
+}
+
+uint64_t RiceBackend::messageSenderDestinationID() const
+{
+    return identifier().toUInt64();
+}
+
+std::optional<SharedPreferencesForWebProcess> RiceBackend::sharedPreferencesForWebProcess() const
+{
+    if (auto connectionToWebProcess = m_connection.get())
+        return connectionToWebProcess->sharedPreferencesForWebProcess();
+
+    return std::nullopt;
+}
+
+GRefPtr<RiceSockets> RiceBackend::getSocketsForStream(unsigned streamId)
+{
+    return m_sockets.get(streamId).sockets;
+}
+
+GRefPtr<GSource> RiceBackend::getRecvSourceForStream(unsigned streamId)
+{
+    return m_sockets.get(streamId).source;
+}
+
+void RiceBackend::notifyIncomingData(unsigned streamId, WebCore::RTCIceProtocol protocol, String&& from, String&& to, std::span<const uint8_t>&& data)
+{
+    callOnMainRunLoopAndWait([&, streamId, protocol, data = WTFMove(data), from = WTFMove(from), to = WTFMove(to)] {
+        messageSenderConnection()->send(Messages::RiceBackendProxy::NotifyIncomingData { streamId, protocol, from, to, data }, messageSenderDestinationID());
```

Comment:
Don't you need a mutable and moves here for the strings

Reply:
mutable yes, but the strings can't be moved (generated constructor for this uses const String&).